proposed/2023/release-staging-and-deprecation.md

```diff
@@ -0,0 +1,86 @@
+
+## Motivation 
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+Many products (Apps / Tools / Frameworks / Utilities / etc. ...) consist of multiple interdependent packages which should be published together as a single set. Currently, NuGet provides no functionality for definding / grouping packages into such sets. This results in package owners having to keep track of these package sets on their own. There are three specific sets of scenarios which this feature will simplify dramatically:
+1. When preparing a release, product owners would like to have a mechanism to pre-stage their product packages, so that the release will consist only of a single action of marking the set of packages as `publishged`, which would result all the packages in the set to be marked as published at the same time.
+2. Sometimes developers have to take a dependency on several related packages in their projects, which are part of the same "release". Without having a concept of the `release` developers have to guess which versions of those packages are compatible. As a result, most of hte time developers would simply pick the latest version of all these packages, which somtimes can lead to incompatibility issues down the road.
+3. Keeping a package healthy on NuGet also means keeping the list of packages aligned with it support policy. Hence, package owners try to deprecate packages which are out of support. Without ability to group packages, package owners end up keeping track of all the packages which belong to the same release and run the deprecation over a discrete set of packages when necessary. Being able to group packages together would enable deprecating a whole release at once, knowing that all the packages which belong to that release will get deprecated.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+There is a new field in the package metadata, which indicates the unique release of the product the package belongs to. This field is called `release` and the value is scoped to the publisher of the package.
+
+When a package is being built, that field in the metadata is set to the same value for all the related packages, that are built together.
+
+Now, with this information in place, new set of NuGet APIs / extensions allow the user to run the following actions:
```

Comment:
A lot of these actions are most appropriate if the published/org is the same and putting that information into the package metadata is IMO too early in the process. 

What if you try to "deprecate" packages that have the same "release", but they're not owned by the same org. 

To eliminate that confusion, how about making this a server only/publisher feature? So you'd be able to push any package as `staging`. Then the server will be able to finesse the UX around it as needed, something that can evolve over time.

Reply:
Well, on the other hand, this information is very similar to the version, which is something known at build time.

As for the following, I don't think this is a valid scenario. Those should be two separate deprecation sessions.
> What if you try to "deprecate" packages that have the same "release", but they're not owned by the same org.



>how about making this a server only/publisher feature? So you'd be able to push any package as staging. Then the server will be able to finesse the UX around it as needed, something that can evolve over time.

Can you please elaborate on this? Are you suggesting that the server should "figure out" the group to assign the staged package to? How would the server do that without any explicit information?

Reply:
The server can generate a batch ID though some API, and nupkgs being uploaded will need to pass that batch ID as an HTTP header when pushing individual nupkgs. Whatever tool does the uploads (if all this grouped release/deprecation features are nuget.org only, I will argue we should then not build it into `dotnet nuget push`) can automate all that complexity when `push --batch *.nupkg` is run.

Your proposal to put the release name in the nuspec allows packages to be uploaded independently more easily. But even with the complexity of the .NET build, my understandings is that an Azure Artifacts feed is created with a GUID or some other generated name, and each part of the build pushes to there. Therefore, I believe it wouldn't be unreasonably difficult to download all the packages from this feed onto a CI agent, and then do the `push *.nupkg` to nuget.org, so that it is done in one transaction.

I think this idea of server generated ID should be described in the document. If it doesn't end up being the accepted design, then under the "alternate designs" section with an explanation of its drawbacks. Otherwise this "package author puts release in nuspec" should be in the alternate design feature.

Honestly, my mind races with security related questions when the package author defines the release id/name. Are different nuget.org accounts allowed to use the same name? Or does the feature enforce each account to have a different "release namespace" per account?  How would that work with something like Azure Artifacts, or [other servers](https://learn.microsoft.com/en-us/nuget/hosting-packages/overview) that implement the nuget server API?  I'm sure there are more risks if we think about it. My gut feeling is that allowing the server to generate the ID makes security easier.

Reply:
The way I was thinking about it is slightly different from @zivkan's. 

My thinking is that we can just have a grouping that's configured in any way shape of work *on* the server itself. 
For example, the server could choose how they represent releases and the packages are configured there. The client *doesn't* need to know that. 

The client can imo know 2 things. 
1. Push as unlisted/staged?
1. Move from staged to "available" or whatever we describe the verb is. 

The client then wouldn't need to know of any new metadata/release id etc. 

IMO by not having this metadata *in* the package itself, we avoid chicken/egg problems of build time vs branding. We don't have any of the ownership questions, since theoretically a `release` could cover multiple owners. 

The disadvantage of this idea is that it'd be impossible to query for these packages locally.

Reply:
Thanks for sharing your thoughts / ideas @zivkan, @nkolev92.
I don't think letting the server decide what the "group" is correct. What if for whatever reason the first batch-upload session got interrupted and now I need to have another session to upload the remaining packages? Or even without interruptions, I just happen to upload the packages which are part of the same release in several sessions.

I just keep coming to the current experience. We do know the version / release the package belongs to when we build it.
Maybe @leecow can share more information about if we ever have a situation, where the branding will be defined after the packages are built?

Reply:
I'm not necessarily referring to the server requiring things to come in one upload. 
It'd be something that can be "connected" independently, regardless of who it was upload/what metadata it has. 

It can be something similar to how API keys are done, for example when you create API keys on the server, you can create scopes for individual package ids. This could work similarly.

Reply:
I think calling it not correct is a bit strong. Your proposal of having the package author name the release whatever they want, and forcing the server to partition the labels by account/whatever is imposing a specific design on server implementations. Maybe that's fine, maybe that has performance risks due to the server's existing implementation. We need input from the server team, and unfortunately we've only had comments from NuGet's client team so far.

In particular, nuget.org's organization accounts add complexity. Say I'm publishing new versions of NuGet.* packages for release 6.8.0. [These packages are owned by the Microsoft org, the NuGet org, and I'd be pushing from my own `zivkan` account](https://www.nuget.org/packages/NuGet.Protocol). Exactly what scope is this `6.8.0` release?

If there's any other package owned by the Microsoft org account with the release label `6.8.0`, do my NuGet.* packages become part of that release?  When I tell nuget.org to "make 6.8.0 public", will I affect those other packages?

Or since I pushed the package, will "release 6.8.0" be scoped to my `zivkan` account. But that means if I stage the packages before GA day, but then I'm sick or no longer an employee, how will someone else publish it as public?  Or mark the whole release as deprecated?

What would be ideal is to be able to say the release is scoped to the NuGet org account, allowing other members of the same org to interact with it, but the Microsoft org can't see this package's release scope, despite being a co-owner of the package.

These are the types of issues I hope the server team can provide good feedback on.

Reply:
> Or since I pushed the package, will "release 6.8.0" be scoped to my zivkan account. But that means if I stage the packages before GA day, but then I'm sick or no longer an employee, how will someone else publish it as public? Or mark the whole release as deprecated?

I think users do things **on behalf** of orgs. So I can "push" a package, but you should be able to change the visibility. If you're not logged in, you can't really see which individual person pushed it.

Reply:
How prescriptive should feature specs be? Is this something that the spec should explicitly mention/address, or leave the spec "open", so different servers can implement it however they see fit, but therefore customers pushing packages have different experiences depending on which server they push to?

Reply:
I agree with much of what @zivkan has said here. There is a decision between a pack-time release label and a server-owned batch/release ID. Both should be considered in full. I think a `batch ID` GUID is one option, but another option is to allow an explicit release entity to be created as a separate gesture and then, over time, packages can be explicitly added or removed to the release entity. This could even be defaulted based on package metadata somewhat marrying the two options.

I think it should be decided now because it has big impact on what the author workflow would be like. Some options require enhancements to `pack`, others require enhancements to `push`, others could be server-only with a `add package to release` or `stage package` gesture that perhaps need not be implemented in NuGet client immediately.